### PR TITLE
Step 2: Setup webpack build for sandpack service worker

### PR DESCRIPTION
**Pull request by souldzin**
- [POC: Sandpack Service Worker to Resolve Relative Files](https://github.com/codesandbox/codesandbox-client/pull/5137)
- [Step 1: Do not register webpack service worker in Sandpack](https://github.com/codesandbox/codesandbox-client/pull/5138)
- **[Step 2: Setup webpack build for sandpack service worker](https://github.com/codesandbox/codesandbox-client/pull/5218)**

## What kind of change does this PR introduce?

This PR sets up webpack to have another entrypoint for building the upcoming Sandpack Service Worker. It's necessary for webpack to build this so that:

- We can have nice typesafety features.
- The service worker and sandpack client can share common types.
- Webpack can add a hash to the service worker so that we can make sure clients have the latest one. Service workers are very eagerly cached by browsers.

This PR also sets up creating a `assets.json` which is a manifest of assets created by webpack. This is important because:

- The entrypoint needs to know where the sandpack service worker lives since the file will have a hash in it to prevent eager caching.
- The Sandpack Service Worker which is going to proxy requests, **does not** proxy requests to things in assets.json (see [the POC for more details](https://github.com/codesandbox/codesandbox-client/pull/5137)).

## What steps did you take to test this? This is required before we can merge, make sure to test the flow you've updated.

There isn't any runtime difference here, and the production assets should be the same except now there are two new files:

- assets.json
- sandpack-service-worker.[hash].js

## Checklist

- [x] Documentation N/A
- [ ] Testing 
- [ ] Ready to be merged (waiting for [Step 1](https://github.com/codesandbox/codesandbox-client/pull/5138) to be merged and this one to be rebased and retargeted)
      
- [ ] Added myself to contributors table

### CODE REVIEW

**Comment by lbogdan** [comment]
Build for latest commit https://github.com/codesandbox/codesandbox-client/pull/5218/commits/332e531aad8131c2232e983b63c0cad2ed6e6295 is at https://pr5218.build.csb.dev/s/new.

File: packages/app/config/setup-entries.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+const AssetsPlugin = require('assets-webpack-plugin');
+const merge = require('webpack-merge');
+const path = require('path');
+const paths = require('./paths');
+
+module.exports = config => {
+  const isProd = config.mode === 'production';
+  const shared = {
+    plugins: [
+      new AssetsPlugin({
```
Comment (souldzin) [inline packages/app/config/setup-entries.js:10]: **Why do we need `AssetsPlugin`?**

See the MR description for more details.

This is a nice plugin which provides a JSON file that can map an entrypoint name to the actual file path that webpack outputs to. So, at runtime I can query for `assets.json` and look for the `sandpack-service-worker` entry to get the hashed file path.

File: packages/app/scripts/build.js [hunk 2]
```diff
@@ -140,7 +145,7 @@ function build(previousSizeMap) {
 
     if (stats.hasWarnings()) {
       console.warn(chalk.yellow('Build warnings:'));
-      stats.compilation.warnings.forEach(({ name, message }) => {
+      getStatsWarnings(stats).forEach(({ name, message }) => {
```
Comment (souldzin) [inline packages/app/scripts/build.js:148]: **What's going on here?**

Now that there are multiple entrypoints, we actually get back a [`MultiStats`](https://webpack.js.org/api/node/#stats-object) object which has the same API, except `compilation` is nested under a `stats` array.

**Comment by souldzin** [comment]
@CompuIves here's the next PR in the Sandpack-Service-Worker-Saga which updates Webpack to build the upcoming service worker (who can see how this is all done in the POC). Check out the PR description for details.

Could you please review and merge? Thanks!

**Review by CompuIves — COMMENTED** [review]
All looks good! I have one question related to the manifest.

File: packages/app/config/setup-entries.js (hunk 1 above)
Comment (souldzin) [reply packages/app/config/setup-entries.js:10]: @CompuIves I think I remember going down this path, but finding some issue with it. But I think you're right! It **should** also do the job. I'm going to take a look 👀

Sorry for the late reply + hope you had a great new year / happy holidays!

**Pull request closed without merging**